File: Dynamic Programming/python/322.py
class Solution:
    def coinChange(self, coins, amount):
        n = len(coins)
        m = amount
        dp = [[0] * (m + 1) for _ in range(n + 1)]
        for i in range(1, m + 1):
            dp[0][i] = float('inf')
        for i in range(1, n + 1):
            for j in range(1, m + 1):
                if coins[i - 1] > j:
                    dp[i][j] = dp[i - 1][j]
                else:
                    remain = j - coins[i - 1]
                    dp[i][j] = min(dp[i - 1][j], 1 + dp[i][remain])
        if dp[n][m] == float('inf'):
            return -1
        return dp[n][m]

File: Dynamic Programming/cpp/322.cpp
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        int n=coins.size();
        int m=amount;
        vector<vector<int>>dp(n+1,vector<int>(m+1,0));
        for(int i=1;i<=m;i++){
            dp[0][i]=1e9;
        }
        for(int i=1;i<=n;i++){
            for(int j=1;j<=m;j++){
                if(coins[i-1]>j){
                    dp[i][j]=dp[i-1][j];
                }
                else{
                    int remain = j-coins[i-1];
                    dp[i][j]= min(dp[i-1][j],1+dp[i][remain]);
                }
            }
        }
        if(dp[n][m]==1e9){
            return -1;
        }
        return  dp[n][m];
    }
};